programs/psa/psa_hash.c

```diff
@@ -0,0 +1,163 @@
+/*
+ *  Example computing a SHA-256 hash using the PSA Crypto API
+ *
+ *  The example computes the SHA-256 hash of a test string using the
+ *  one-shot API call psa_hash_compute() and the using multi-part
+ *  operation, which requires psa_hash_setup(), psa_hash_update() and
+ *  psa_hash_finish(). The multi-part operation is popular on embedded
+ *  devices where a rolling hash needs to be computed.
+ *
+ *
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+
+#include "psa/crypto.h"
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "mbedtls/build_info.h"
+#include "mbedtls/platform.h"
+
+#define HASH_ALG PSA_ALG_SHA_256
+
+#define TEST_SHA256_HASH {                                                 \
+        0x5a, 0x09, 0xe8, 0xfa, 0x9c, 0x77, 0x80, 0x7b, 0x24, 0xe9, 0x9c, 0x9c, \
+        0xf9, 0x99, 0xde, 0xbf, 0xad, 0x84, 0x41, 0xe2, 0x69, 0xeb, 0x96, 0x0e, \
+        0x20, 0x1f, 0x61, 0xfc, 0x3d, 0xe2, 0x0d, 0x5a                          \
+}
+
+const uint8_t mbedtls_test_sha256_hash[] = TEST_SHA256_HASH;
+
+const size_t mbedtls_test_sha256_hash_len =
+    sizeof(mbedtls_test_sha256_hash);
+
+#if !defined(MBEDTLS_PSA_CRYPTO_C) || !defined(PSA_WANT_ALG_SHA_256)
+int main(void)
+{
+    mbedtls_printf("MBEDTLS_PSA_CRYPTO_C and MBEDTLS_SHA256_C"
+                   "not defined.\r\n");
+    return EXIT_SUCCESS;
+}
+#else
+
+int main(void)
+{
+    uint8_t buf[] = "Hello World!";
```

Comment:
`buf` includes a terminating null byte. We need to either not include it, or use `strlen(buf)` or `sizeof(buf) - 1` as the message length. Calculating the hash of a message with a terminating null byte is technically correct, but is not human-friendly and so is not suitable for a demo.

Reply:
Slightly confused by this comment. I understand about terminating null bytes but do you mean that we should be using `sizeof(buf) -1` in the calls like this below?
`status = psa_hash_update(&hash_operation, buf, sizeof(buf));`

Reply:
Yes, either
```
status = psa_hash_update(&hash_operation, buf, sizeof(buf) - 1);
```
or
```
status = psa_hash_update(&hash_operation, buf, strlen(buf));
```
Or alternatively arrange for `buf` to not contain a trailing null byte, but that's not convenient since we want to use a string literal.